### ISSUE TITLE
Do not accept any connections when high memory usage is detected

### ISSUE THREAD

**Issue by mariusstaicu** [emqx/emqx#11891]
### What would you like to be added or enhanced?

It would be nice if there was a configuration to stop accepting new connections when a certain memory threshold is reached.

### Why is this needed?

This would help a lot with stability when one of the brokers in a cluster restarts discarding all its connected clients. Then a wave of connections flow to other brokers which could lead to cascading broker restarts because of out of memory.

By not accepting new connections when a broker is, for example, using more than 70-75% of his memory, would make that broker impervious to what happens in the cluster.

For example, Elasticsearch has implemented this in two thresolds:
- low disk watermark - when this is reached, the node stops accepting new data
- high disk watermark - when this is reached, the node tries to move data to other brokers so it can free some disk space

Emqx could to something analogue to this, on the memory side.

**Comment by lafirest** [comment]
@mariusstaicu 
We already have a similar feature https://github.com/emqx/emqx/blob/03358b77aa571e083ae598c42a844c7293a4f551/apps/emqx/src/emqx_olp.erl#L89, regrettably is the trigger condition not dependent on the memory.

**Comment by qzhuyan** [comment]
note: 

we didn't do it since 

1. The VM memory usage does not reflect what the beam process mem consumption is using in OS due to our NIF threads, mem fragmentation, and caches. So basically we must read from the OS/platform.


1. We have an existing impl. of reading the meminfo from the platform for alerting. 
for the emqx running in the container, we sometimes cannot read the proper mem limit of the process in the platform.
for example, the container is configured to max use 2GB mem but when EMQX reads the meminfo via the platform interface, it get the actual meminfo of the whole host OS. 
The behavior various in different platforms, kernel versions or container orchestrations. 


however, I think it is worth to impl this  feature  by
1. read the meminfo of the running process from the platform, not using the old API which reads the meminfo of the platform.
2. intro extra config switch that turns on/off this feature [backoff when lowmem]. 0 means off,  >0 is the threshold. 
3. into extra config of what is the max mem EMQX is allowed to use (assume we cannot get the system limit)

**Comment by mariusstaicu** [comment]
Yes, indeed, I noticed that for some pods the dashboard is able to show emqx mem usage / container mem limit, but for some pods it shows the entire k8s node memory usage:
![emqx-node-mem](https://github.com/emqx/emqx/assets/9932238/ea80f28b-d46e-4533-9b68-cb21171a0b43)
